File: lib/core/services/scan_pipeline.dart
import 'dart:io';
import 'package:uuid/uuid.dart';

import 'fuzzy_matcher.dart';
import 'knowledge_base.dart';
import 'ml_engine.dart';
import 'ocr_service.dart';
import 'parsers.dart';

const _uuid = Uuid();

/// Orchestrates: OCR → parse → match → explain → predict grade.
/// Runs entirely on-device, no network needed.
class ScanPipeline {
  static final ScanPipeline instance = ScanPipeline._();
  ScanPipeline._();

  FuzzyMatcher? _matcher;
  bool _ready = false;

  Future<void> init() async {
    if (_ready) return;
    await MlEngine.instance.load();
    await KnowledgeBase.instance.load();
    _matcher = FuzzyMatcher(KnowledgeBase.instance.vocabList, cutoff: 70);
    _ready = true;
  }

  Future<Map<String, dynamic>> scan(File imageFile) async {
    if (!_ready) await init();

    // 1. OCR
    final ocr = await OcrService.instance.recognize(imageFile);
    if (ocr.error != null) {
      return {'error': ocr.error, 'scan_id': _uuid.v4()};
    }

    // 2. Panel type detection
    final panelType = detectPanelType(ocr.rawText);

    List<Map<String, dynamic>> matched = [];
    List<String> unmatched = [];
    List<String> eNumbers = [];
    List<Map<String, dynamic>> additives = [];
    Map<String, double> nutrients = {};

    // 3. Ingredients branch
    if (panelType == 'ingredients' || panelType == 'both') {
      final parsed = parseIngredients(ocr.rawText);
      eNumbers = parsed.eNumbers;

      // Fuzzy match tokens to vocabulary
      for (final token in parsed.tokens) {
        final result = _matcher!.bestMatch(token);
        if (result != null) {
          final info =
              KnowledgeBase.instance.getIngredientInfo(result.match);
          matched.add({
            'ocr_text': token,
            'matched_to': result.match,
            'confidence': result.score.round(),
            if (info != null) 'explanation': info['what'],
            if (info != null && info['safety'] != null)
              'safety': info['safety'],
          });
        } else {
          unmatched.add(token);
        }
      }

      // Explain additives
      for (final code in eNumbers) {
        final info = KnowledgeBase.instance.getAdditiveInfo(code);
        if (info != null) {
          additives.add(info);
        } else {
          additives.add({
            'code': code,
            'name': 'Unknown Additive',
            'what': 'This additive code was not found in our database.',
            'safety': 'Check official food safety resources.',
          });
        }
      }
    }

    // 4. Nutrition branch
    if (panelType == 'nutrition' ||
        panelType == 'both' ||
        panelType == 'unknown') {
      nutrients = parseNutrition(ocr.rawText);
    }

    // 5. Grade prediction
    final prediction = MlEngine.instance.predict(
      nutrients,
      additivesCount: eNumbers.length,
    );

    // Low confidence guard
    if (nutrients.length < 3 &&
        (panelType == 'nutrition' || panelType == 'unknown')) {
      prediction['grade_label'] =
          '${prediction['grade_label']} (LOW CONFIDENCE — few nutrients readable)';
    }

    final scanId = _uuid.v4();

    return {
      'scan_id': scanId,
      'panel_type': panelType,
      'raw_ocr': ocr.rawText,
      'ocr_confidence': ocr.meanConfidence,
      'nutrients': nutrients,
      'ingredients': matched,
      'unmatched': unmatched,
      'additives': additives,
      'ingredient_count': matched.length,
      'created_at': DateTime.now().toIso8601String(),
      ...prediction,
    };
  }
}

File: nutriscan/backend/app/services/scan_pipeline.py
"""
Scan orchestrator — the production equivalent of the notebook's scan_food_packet().
OCR → detect panel → parse → match → explain → predict grade.
"""
from sqlalchemy.orm import Session

from app.core.logging import get_logger
from app.services.ml.predictor import ModelService
from app.services.ocr.engine import run_ocr
from app.services.ocr.parsers import (detect_panel_type, parse_ingredients,
                                      parse_nutrition)
from app.services.rag.explainer import ExplanationService

logger = get_logger("pipeline")


async def run_scan_pipeline(image_bytes: bytes, db: Session) -> dict:
    ml = ModelService.get()
    explainer = ExplanationService(db)

    # 1. OCR
    ocr = run_ocr(image_bytes)
    if ocr.error:
        return {"error": ocr.error}

    # 2. Panel type
    panel_type = detect_panel_type(ocr.raw_text)

    matched: list[dict] = []
    unmatched: list[str] = []
    e_numbers: list[str] = []
    additives: list[dict] = []
    nutrients: dict[str, float] = {}

    # 3. Ingredients branch
    if panel_type in ("ingredients", "both"):
        tokens, e_numbers = parse_ingredients(ocr.raw_text)
        matched, unmatched = ml.match_ingredients(tokens)

        # explain matched ingredients (DB → LLM cascade)
        for m in matched:
            exp = await explainer.explain(m["matched_to"])
            if exp:
                m["explanation"] = exp["what"]
                m["safety"] = exp.get("safety")

        # explain additives (curated dict → KB frequency fallback)
        for code in e_numbers:
            info = explainer.curated_additive(code)
            if info is None:
                kb = ml.additives_kb
                row = kb[kb["code"] == code.upper()]
                if not row.empty:
                    info = {"code": code.upper(), "name": "Additive",
                            "what": f"Found in {int(row.iloc[0]['count']):,} products in our database.",
                            "safety": f"Most commonly appears in grade '{row.iloc[0]['most_common_grade']}' products.",
                            "source": "kb_frequency"}
            if info:
                additives.append(info)

    # 4. Nutrition branch
    if panel_type in ("nutrition", "both", "unknown"):
        nutrients = parse_nutrition(ocr.raw_text)

    # 5. Grade
    prediction = ml.predict_grade(nutrients, additives_count=len(e_numbers))

    # Low information guard: if we extracted almost nothing, don't pretend confidence
    if len(nutrients) < 3 and panel_type in ("nutrition", "unknown"):
        prediction["grade_label"] = prediction["grade_label"] + " (LOW CONFIDENCE — few nutrients readable)"

    logger.info("scan_complete", panel_type=panel_type,
                ingredients=len(matched), additives=len(e_numbers),
                nutrients=len(nutrients), grade=prediction["grade"],
                ocr_confidence=ocr.mean_confidence)

    return {
        "panel_type": panel_type,
        "raw_ocr": ocr.raw_text,
        "ocr_confidence": ocr.mean_confidence,
        "nutrients": nutrients,
        "ingredients": matched,
        "unmatched": unmatched,
        "additives": additives,
        **prediction,
    }
